services/sharedchannel/sync_send.go

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package sharedchannel
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/mattermost/go-i18n/i18n"
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/services/remotecluster"
+	"github.com/mattermost/mattermost-server/v5/utils"
+	"github.com/wiggin77/merror"
+)
+
+type syncTask struct {
+	id         string
+	channelId  string
+	remoteId   string
+	AddedAt    time.Time
+	retryCount int
+}
+
+func newSyncTask(channelId string, remoteId string) syncTask {
+	return syncTask{
+		id:        channelId + remoteId, // combination of channelId and remoteId to avoid duplicates
```

Comment:
This id allow us to serialize sync on a per channel per remote basis where message ordering matters.  I may change this to serialize per remote now that we are sync'ing users as well.

That said, before the holidays I noticed a flaw in the threading model where one thread could pop the next message for a remote before the previous one is finished on another thread.  I talked to George about it and we agreed I'd defer fixing that for now as I'll just limit the thread pool to one.  Fine for immediate use cases.

Reply:
My worry was about possibly losing a task in `addTask` since we only set a task if there isn't one with the same id already. Couldn't more than one task for the same channel and remote pair be created? E.g. two close in time sync events.

Reply:
The "task" is just a signal that a remote needs fresh data from the channel.  Once the task is popped, it then fetches the fresh data.  If the channel changes while the data is in flight, that new task will enqueue.  There is no need to allow multiple tasks for the same remote/channel... in fact I wanted to avoid that so we don't choke the queue with redundant tasks for a busy channel.

Reply:
Thank you, that makes total sense :+1: